apps/bbsync/query.py

```diff
@@ -307,3 +312,30 @@ def generate_srt_notes(self):
         """
         srt_notes_builder = SRTNotesBuilder(self.flaw, self.old_flaw)
         self._query["cf_srtnotes"] = srt_notes_builder.content
+
+    def generate_comment(self):
+        pending_comments = FlawComment.objects.pending().filter(flaw=self.flaw)
+
+        # These checks are not triggerable with ATOMIC_REQUESTS, but
+        # are present as a failsafe because posting old pending comments
+        # might have bad consequences.
+        if pending_comments.count() > 1:
+            raise BadRequest(
+                "Cannot add more than one comment with one update! "
+                "New comment requests running in parallel on the same flaw."
```

Comment:
Thank you for raising a warning about this. After some preliminary investigation with submitting concurrent requests, I see that these checks are insufficient to prevent posting multiple comments with the same `order`. The experiment resulted in two conflicting comments actually ending up being saved in bugzilla and all other requests resulting in uncleaned temporary `FlawComment` objects, which borked the OSIDB instance (and prevented future flaw syncs) until I deleted them manually from the DB.

 It seems that `ATOMIC_REQUESTS` in itself is insufficient. The reason is most probably the default `isolation_level` is happy to run concurrent operations even if each operation needs to have the most up-to-date read view of the database. I yet have to find out why ATOMIC_REQUESTS sufficiently fails a single non-concurrent request that gets a 40x from bugzilla (and cleans up the temporary FlawComment instance), but not when done in parallel.

I'll investigate whether there's a way to signal to Django that the given view method must not run concurrently (a lock? a different isolation_level?) and make sure a bugzilla 40x makes the osidb request fail and the database properly rolled back. Any pointers, please?

This should make it possible to remove the checks you commented on.

Reply:
Interesting analysis! I would say that we should balance the safeness and the complexity. It is not such a big deal if there are two concurrent comments really heading to Bugzilla without properly catching the collision. What would be a serious issue however is if OSIDB got in some blocked state when no one can add comments.

We can pretty much (probably not totally perfectly) mitigate this on our API as the potential collision time between hitting the API and checking the comment order in the DB should be very small. It is harder between OSIDB and Bugzilla where the minimum collision window will be maybe 1s all the time. I would just accept it and not go crazy about it :smile:

The real probability of two concurrent comments is then very small so I think an acceptable risk and eventual duplicite comment is not such a big deal.

Reply:
Resolved in https://github.com/RedHatProductSecurity/osidb/pull/209/commits/f9ebfa5badd141532da083c3d6506b796e4868b4